Show Caller Name in XREF

lorsanta opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [x] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [ ] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)


**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Ref #3416

**Test plan (required)**

Open a project, select a function and check for xref.
The dialogue now can also show the name of the caller.

![2025-02-26 23-49-08](https://github.com/user-attachments/assets/bb8dc6c8-24b0-45a5-941e-4a6864a1da3c)

<!-- What steps should the reviewer take to test your pull request? Demonstrate that the code is solid. Example: The exact actions you made and their outcome. Add screenshots/videos if the pull request changes UI. This is your time to re-check that everything works and that you covered all the edge cases -->


<!-- **Code formatting**
Make sure you ran clang-format on your code before making the PR. Check our contribution guidelines here: https://cutter.re/docs/contributing/code/getting-started.html -->

**Closing issues**

Closes #3416

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if such). -->

wargio reviewed (approved):
Looks good! @karliss

karliss commented:
Rizin should already have a function which formats an address  as "name+offset" (can't comment whether there is a suitable Cutter wrapper for it already), where name might be a function and at least in theory it could be something else like a global array name if xref is done by a pointer in some kind of table.
Other benefit is that it will make things a bit clearer when a function contains multiple xrefs to the same thing.




For example here is backtrace widget doing similar thing:
![image](https://github.com/user-attachments/assets/9d76bbdf-4c7e-403a-90bd-a149a1a07c27)
And here is disassembly with `asm.reloff`  option turned on.
![image](https://github.com/user-attachments/assets/9a38cdb4-1719-4e54-885f-d4bde86968d8)

Usefulness for "Xrefs from" seems very limited. For all the call instructions the target function should be already visible, possibly twice. Once in address column and second time in code column. No point adding extra noise by displaying same information third time.
![image](https://github.com/user-attachments/assets/31841ea6-d78b-48f3-b3de-3ed55ebee5ed)

wargio reviewed (changes requested):
Hmm makes sense. this is true also for xrefs that are not calls. (like data xrefs etc..), where there is "no caller"

notxvilka commented:
Yes, there is a function: `RzFlagItem *f = rz_flag_get_at(core->flags, addr, !strict_offset);` It's used in the Rizin's `fd` command that returns precisely this: name + offset

lorsanta commented:
I didn't found a specific function which given an address formats it as "name+offset", rather it seems it's copy pasted in different parts in Rizin: 
https://github.com/rizinorg/rizin/blob/58a26a749487568f088cd8b71c4d0d307675fbc9/librz/core/disasm.c#L2787-L2791
https://github.com/rizinorg/rizin/blob/58a26a749487568f088cd8b71c4d0d307675fbc9/librz/core/cmd/cmd_flag.c#L348-L351
https://github.com/rizinorg/rizin/blob/58a26a749487568f088cd8b71c4d0d307675fbc9/librz/core/cdebug.c#L921-L928

If I can't do it the way @notxvilka suggests it, I'm not sure what I could do.

wargio commented:
The last link you have shared does exactly what you have described. flag + offset (just called delta there)

lorsanta commented:
Sorry, I'm confused @wargio , I could do something like this:
```diff --git a/src/core/Cutter.cpp b/src/core/Cutter.cpp
index e674d9a6..3abd9b21 100644
--- a/src/core/Cutter.cpp
+++ b/src/core/Cutter.cpp
@@ -4021,7 +4021,24 @@ QList<XrefDescription> CutterCore::getXRefs(RVA addr, bool to, bool whole_functi
             continue;
         }
 
-        xd.from_str = RzAddressString(xd.from);
+        if(to) {
+            RzFlagItem *f = rz_flag_get_at(Core()->core()->flags, xd.from, true);
+            if(f) {
+                int delta = xd.from - f->offset;
+                if (delta > 0) {
+                    xd.from_str = rz_str_newf("%s+%x", f->name, delta);
+                } else if (delta < 0) {
+                    xd.from_str = rz_str_newf("%s-%x", f->name, delta);
+                } else {
+                    xd.from_str = rz_str_newf("%s", f->name);
+                }
+            } else {
+                xd.from_str = RzAddressString(xd.from);
+            }
+        }
+        else {
+            xd.from_str = RzAddressString(xd.from);
+        }
         xd.to_str = Core()->flagAt(xd.to); 
```
But wouldn't that contradict what @karliss  was saying about not reimplementing things on cutter side.

wargio commented:
yes, but delta is already an `int` so you should probably use `%d` instead of `%x`

and modify the negative to not include the sign, which will be added by `rz_str_newf`
```
+                } else if (delta < 0) {
+                    xd.from_str = rz_str_newf("%s%d", f->name, delta);
```

wargio commented:
there seems to be an issue with jsdec (unrelated to this PR), let me fix it and we can restart the failing jobs and probably merge.

notxvilka commented:
I wonder if it makes sense to make it an API function from the Rizin side. To avoid copy-pasting the same code between Rizin and cutter. Something like `rz_flag_describe()`

wargio commented:
yes probably we can merge this and then have an api returning a string.

wargio reviewed (approved):
LGTM

wargio commented:
@karliss give it a look

karliss commented on src/core/Cutter.cpp:
```diff
@@ -4021,7 +4021,23 @@ QList<XrefDescription> CutterCore::getXRefs(RVA addr, bool to, bool whole_functi
 
-        xd.from_str = RzAddressString(xd.from);
+        if (to) {
+            RzFlagItem *f = rz_flag_get_at(Core()->core()->flags, xd.from, true);
```
`foo(Core()->core()->bar)` is almost always going to be mistake especially if foo return a non self contained object (ie something owned by rizin).  Move the `CORE_LOCK();` used by rz_analysis_xrefs_get one scope higher and use that.

karliss reviewed (changes requested):
As a side note I highly recommend to find a suitable resource and learn the basics of git. Making pull requests from the dev branch of your fork will likely cause you problems in future.

lorsanta commented:
Sorry for the mess but I think I'm finally on the right track: I replaced `rz_str_newf()` with `QString::arg()` (didn't used `QString::asprintf()` since the qt docs recommend not to use it), and moved `CORE_LOCK();` one scope higher

karliss reviewed (changes requested):
The code is fine. 

But I have doubts about  merging it in current state.

It adds a bunch of code which will have to be later thrown out, and contains no code which will be useful later once rizin has appropriate API. Sure there are situations where it's useful  where it's useful to merge partial implementation first, and then replace some of it withe calls to better API when available.  But that assumes that the changes contain at least some useful code (some UI additions or code restructuring) which will be reusable once the better API is available. 

Other case where it might be worth merging partial implementation is if it still provides significant improvement in usability. That in current state is also limited due to inconsistency with rizin and not handling a bunch of cases which rizin does. 

Here are some examples:
Useless "case.xxx.yy+offset" since the current version takes any flag without considering it's type.
![image](https://github.com/user-attachments/assets/fd0e1829-f21a-4dd7-815d-df5ec6cabcfe)
Uses different style of formatting `fcn.00b200+81` vs `fcn.00b200+0x51`.
![image](https://github.com/user-attachments/assets/cdf856f1-b282-49a4-9c5b-32946ac2a589)
I have also seen cases where current implementation displays `data_something+offset` even though asm.reloff displayed `fcn.xxxx+offset`. And that was with `asm.relof.flags` also enabled.


For the temporary implementation might be better to use `CutterCore::functionIn`  like the initial implementation did (but keeping rest of the structure from current version).  While in theory this feature should display of not just function relative stuff, in the current version use of `rz_flag_get_at`  makes the result worse more often than it improves. Need to try out with functionIn to be sure.

Also should use offset as ~hexadecimal~ number like the `asm.reloff` mode does.  

Correction:
`asm.reloff` chooses base based on `asm.decoff`
